packages/syntax/src/renderingPlugin.tsx

```diff
@@ -0,0 +1,177 @@
+
+  const numberCellStyle = css`
+    padding-right: 24px;
+    color: ${numberColor};
+  `;
+
+  return (
+    <tr>
+      <td className={cx(cellStyle, numberCellStyle)}>{lineNumber}</td>
+      <td className={cellStyle}>{children}</td>
+    </tr>
+  );
+}
+
+export function treeToLines(
+  children: Array<string | TokenObject>,
+): Array<Array<TreeItem>> {
+  const lines: Array<Array<TreeItem>> = [];
+  let currentLineIndex = 0;
+
+  // Create a new line, if no lines exist yet
+  if (lines[currentLineIndex] == null) {
+    lines[currentLineIndex] = [];
+  }
+
+  children.forEach(child => {
+    if (isString(child)) {
+      // If the current element is a string that includes a line break, we need to handle it differently
+      if (child.includes('\n')) {
+        child.split('').forEach(fragment => {
+          if (fragment === '\n') {
+            // If the fragment is a new line character, we create a new line
+            currentLineIndex++;
+            lines[currentLineIndex] = [];
+          } else {
+            const currentIndexInLine = lines[currentLineIndex].length - 1;
+
+            if (isString(lines[currentLineIndex][currentIndexInLine])) {
+              // If the last element in the line is a string, we append this string to it
+              lines[currentLineIndex][currentIndexInLine] += fragment;
+            } else {
+              // Otherwise, we push the string fragment on its own
+              lines[currentLineIndex].push(fragment);
+            }
+          }
+        });
+      } else {
+        // We don't need to do anything special in the case where the string doesn't contain a line break
+        lines[currentLineIndex].push(child);
+      }
+    }
+
+    // Line breaks aren't a part of token objects, so we can assume those objects go on the current line
+    if (isObject(child)) {
+      lines[currentLineIndex].push(child);
+    }
+  });
+
+  // Strip empty lines from the beginning of code blocks
+  while (lines[0].length === 0) {
+    lines.shift();
+  }
+
+  // Strip empty lines from the end of code blocks
+  while (lines[lines.length - 1].length === 0) {
+    lines.pop();
+  }
+
+  return lines;
+}
+
+function renderLineAsTableRow(line: Array<TreeItem>, index: number) {
+  return (
+    <LineTableRow key={index} lineNumber={index + 1}>
+      {line.map(processToken)}
+    </LineTableRow>
+  );
+}
+
+function renderLineAsFragment(line: Array<TreeItem>) {
+  return (
+    <>
+      <span>{line.map(processToken)}</span>
+
+      {/* We use a new line character here instead of <br /> so that text will not break outside of a <pre /> tag */}
+      {'\n'}
+    </>
+  );
+}
+
+const plugin: HighlightPluginEventCallbacks = {
+  'after:highlight': function (result) {
+    const { rootNode } = result.emitter;
+    const lines = treeToLines(rootNode.children);
+
+    result.react = lines.map(renderLineAsFragment);
+    result.reactWithNumbers = lines.map(renderLineAsTableRow);
```

Comment:
performance nit, you can loop over lines once and do the assignment inside there instead.

Reply:
Excellent point – will do!